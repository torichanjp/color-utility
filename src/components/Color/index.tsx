import classes from "./Color.module.sass"
import {Color as ColorElement, ColorArray} from "../../lib/color";
import React from "react";
import {useParams} from "../../contexts/ParamsContext";

export const Color: React.FC<{color: ColorElement, text?: string}> = (props) => {
    const {color, text} = props
    const backgroundColor = '#' + color.hex
    const textColor = color.luma >= 55 ? 'black' : 'white'
    const textJsx = (text
        ? [text]
        : [
            '#' + color.hex,
            `Hue: ${color.hsv.h}`,
            `Sat: ${color.hsv.s}`,
            `Val: ${color.hsv.v}`,
            `Luma: ${color.luma}`
        ])
        .map((_text, i) => {
            return (
                <p key={i}>{_text}</p>
            )
    })

    return (
        <div
            aria-label={color.hex}
            className={classes.color__item}
            style={{backgroundColor: backgroundColor, color: textColor}}>
            {textJsx}
        </div>
    )
}

export const Colors: React.FC<{colors: ColorElement[]}> = (props) => {
    const {colors} = props
    const colorsTsx = colors.map((color) => {
        return (
            <Color key={color.hex} color={color}/>
        )
    })
    return (
        <div className={classes.color__container}>
            {colorsTsx}
        </div>
    )
}

export const ColorsWithParams: React.FC<{colors: ColorArray}> = (props) => {
    const { colors } = props
    const { useLuma, hueDivisions} = useParams()

    return (
        <>
            <Colors colors={colors.sortDefault(hueDivisions, useLuma).reverse()}/>
        </>
    )
}
