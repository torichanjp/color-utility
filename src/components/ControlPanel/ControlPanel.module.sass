@use "../../lib/variables" as var

.wrapper
  display: flex
  flex-direction: row
  flex-wrap: wrap
  padding: 1em
  background-color: var.$background-color
  div.firstSortKey
    color: white
    margin: .1em .5em
    p
      margin: 0 0 .2em 0
      &.description
        font-size: smaller
  div.secondSortKey
    color: white
    margin: .1em .5em
    p
      margin: 0 0 .2em 0
    label
      padding: .2em
  button
    padding: .5em
    font-size: large
    min-width: 5em
  input[type=text]
    padding: .5em
    margin: 0 1em
    font-size: large
