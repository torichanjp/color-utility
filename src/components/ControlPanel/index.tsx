import React, { useContext } from "react";
import classes from "./ControlPanel.module.sass"
import { useParams } from "../../contexts/ParamsContext"

export const ControlPanel: React.FC = () => {
    const { useLuma, hueDivisions, setHueDivisions, setUseLuma } = useParams()

    const buttonClick = (value: number, minusPlus: number, condition: Function, minMax: number) => {
        const n = value + minusPlus
        const v = condition(n) ? minMax : n
        setHueDivisions(v)
    }
    const minus1ButtonClick = () => {
        buttonClick(hueDivisions, -1, (n: number) => n <= 0, 0)
    }
    const minus10ButtonClick = () => {
        buttonClick(hueDivisions, -10, (n: number) => n <= 0, 0)
    }
    const plus1ButtonClick = () => {
        buttonClick(hueDivisions, +1, (n: number) => n >= 360, 360)
    }
    const plus10ButtonClick = () => {
        buttonClick(hueDivisions, +10, (n: number) => n >= 360, 360)
    }
    const valueChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
        const v = parseInt(e.target.value)
        if (v >= 0) {
            setHueDivisions(v > 360 ? 360 : v)
        }
    }
    const checkboxChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
        setUseLuma(e.target.checked)
    };
    return (
        <div className={classes.wrapper}>
            <button onClick={minus10ButtonClick}>-10</button>
            <button onClick={minus1ButtonClick}>-1</button>
            <input type="text" value={hueDivisions} onChange={valueChanged}/>
            <button onClick={plus1ButtonClick}>+1</button>
            <button onClick={plus10ButtonClick}>+10</button>
            <div>
                <input id="use-luma" type={"checkbox"} checked={useLuma} onChange={checkboxChanged}/>
                <label htmlFor={"use-luma"}>Use luma</label>
            </div>

        </div>
    )
}
