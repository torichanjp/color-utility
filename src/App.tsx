import React from 'react';
import './App.css';
import { ColorsWithParams } from "./components/Color"
import { ColorArray, Color } from "./lib/color";
import { ControlPanel } from "./components/ControlPanel";
import { ParamsContext } from "./contexts/ParamsContext"
import {colorObjs} from "./data/data";

function App() {
    const colorElements = colorObjs.map(color => new Color(color))
    const colorArrays = new ColorArray(...colorElements)

    return (
        <div className="App">
            <ParamsContext>
                <ControlPanel/>
                <ColorsWithParams colors={colorArrays}/>
            </ParamsContext>
        </div>
    );
}

export default App;
