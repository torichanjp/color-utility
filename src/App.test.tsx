import React from 'react';
import { render, screen } from '@testing-library/react';
import App from './App';
import { colorObjs } from "./data/data";
import {ControlPanel} from "./components/ControlPanel";

const nColor = colorObjs.length

test('renders some elements', () => {
  render(<App />);
  const buttons = screen.queryAllByRole("button");
  expect(buttons).toHaveLength(4)

  // all colors
  const divs = screen.queryAllByText(/#[a-f0-9]{6}/i)
  expect(divs).toHaveLength(nColor)
});

test("Snap shot for App", () => {
  const view = render(<ControlPanel />);
  expect(view.container).toMatchSnapshot();
});
