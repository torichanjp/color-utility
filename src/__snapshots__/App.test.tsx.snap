// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snap shot for App 1`] = `
<div>
  <div
    class="wrapper"
  >
    <button>
      -10
    </button>
    <button>
      -1
    </button>
    <input
      aria-label="num-of-hue-divisions"
      type="text"
      value="10"
    />
    <button>
      +1
    </button>
    <button>
      +10
    </button>
    <div>
      <input
        checked=""
        id="use-luma"
        type="checkbox"
      />
      <label
        for="use-luma"
      >
        Use luma
      </label>
    </div>
  </div>
</div>
`;
